provider/pihole/clientV6.go

```diff
@@ -0,0 +1,376 @@
+
+	// Get the current record
+	if strings.Contains(ep.DNSName, "*") {
+		return provider.NewSoftError(errors.New("UNSUPPORTED: Pihole DNS names cannot return wildcard"))
+	}
+
+	switch ep.RecordType {
+	case endpoint.RecordTypeA, endpoint.RecordTypeAAAA:
+		apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s %s", ep.Targets, ep.DNSName)))
+	case endpoint.RecordTypeCNAME:
+		if ep.RecordTTL.IsConfigured() {
+			apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s,%s,%d", ep.DNSName, ep.Targets, ep.RecordTTL)))
+		} else {
+			apiUrl = fmt.Sprintf("%s/%s", apiUrl, url.PathEscape(fmt.Sprintf("%s,%s", ep.DNSName, ep.Targets)))
+		}
+	}
+
+	req, err := http.NewRequestWithContext(ctx, action, apiUrl, nil)
+	if err != nil {
+		return err
+	}
+
+	_, err = p.do(req)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *piholeClientV6) retrieveNewToken(ctx context.Context) error {
+	if p.cfg.Password == "" {
+		return nil
+	}
+
+	apiUrl := fmt.Sprintf("%s/api/auth", p.cfg.Server)
+	log.Debugf("Fetching new token from %s", apiUrl)
+
+	// Define the JSON payload
+	jsonData := []byte(`{"password":"` + p.cfg.Password + `"}`)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, apiUrl, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return err
+	}
+
+	jRes, err := p.do(req)
+	if err != nil {
+		return err
+	}
+
+	// Parse JSON response
+	var apiResponse ApiAuthResponse
+	err = json.Unmarshal(jRes, &apiResponse)
+	if err != nil {
+		fmt.Println("Error reading response:", err)
+	} else {
+		// Set the token
+		if apiResponse.Session.SID != "" {
+			p.token = apiResponse.Session.SID
+		}
+	}
+	return err
+}
+
+func (p *piholeClientV6) checkTokenValidity(ctx context.Context) (bool, error) {
+	if p.token == "" {
+		return false, nil
+	}
+
+	apiUrl := fmt.Sprintf("%s/api/auth", p.cfg.Server)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, apiUrl, nil)
+	if err != nil {
+		return false, nil
+	}
+
+	jRes, err := p.do(req)
+	if err != nil {
+		return false, nil
+	}
+
+	// Parse JSON response
+	var apiResponse ApiAuthResponse
+	err = json.Unmarshal(jRes, &apiResponse)
+	if err != nil {
+		fmt.Println("Error reading response:", err)
+		return false, err
+	}
+	return apiResponse.Session.Valid, nil
+}
+
+func (p *piholeClientV6) do(req *http.Request) ([]byte, error) {
+	req.Header.Add("content-type", "application/json")
+	if p.token != "" {
+		req.Header.Add("X-FTL-SID", p.token)
+	}
+	res, err := p.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	jRes, err := io.ReadAll(res.Body)
+	defer res.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK &&
+		res.StatusCode != http.StatusCreated &&
+		res.StatusCode != http.StatusNoContent {
+		// Parse JSON response
+		var apiError ApiErrorResponse
+		err := json.Unmarshal(jRes, &apiError)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
if err := json.Unmarshal(jRes, &apiError); err != nil {
	return nil, fmt.Errorf("failed to unmarshal error response: %w", err)
}